### DISCUSSION TITLE
Add a dry-run publish API

### DISCUSSION THREAD

**Discussion by joelgallant** [rust-lang/crates.io#4329] [Feature Requests]
In https://github.com/rust-lang/cargo/pull/6150, @sfackler suggested that crates.io could have a dry-run publishing API endpoint, so it can enforce any requirements (like wildcard deps) in `cargo publish --dry-run`.

I'd love to help out with that, if it's something worth doing.

**Comment by nkconnor** [comment]
Reserved names would also be useful to catch with this; e.g. I got hit with

> error: api errors (status 200 OK): cannot upload a crate with a reserved name

trying to publish my first crate, and had to dig pretty hard to find that it was part of the [DB migration here ](https://github.com/rust-lang/crates.io/blob/a27c704faa2982ddd75a3dc564da85c0217b950e/migrations/20170305095748_create_reserved_crate_names/up.sql#L6)

**Comment by jtgeibel** [comment]
If anyone has time to dig into this, now might be a good time to revisit this. We've been tracking some refactoring in #1522 and this may be easier to implement now. The goal is to make sure the desired checks are done early in the pipeline, before writing anything to the database. All database writes should occur after we return from the `--dry-run` publish and the only external service I think we should possibly hit is to check team membership if necessary.

**Comment by wiktor-k** [comment]
Great idea! Currently `cargo publish --dry-run` runs only client checks and since crates.io has additional policies (e.g. max number of keywords or certain fields being present) it would be good to have that endpoint to validate that *everything* looks OK.

Ideally it'd be offered without authentication so the check can run in CI without tokens or credentials.

Alternatively I think crates.io should have an endpoint describing in machine-readable way what are the current policies. What has been suggested in this discussion is easier to implement, though, so big :+1: from me!
